_posts/2018/2018-04-01-stop-robots-crawlers-triggering-errors-rails.md

```diff
@@ -0,0 +1,106 @@
+---
+
+When a Rails application can't find a record, it throws a 404 error. This is a standard HTTP code for browsers (and other clients) meaning 'not found'.
+
+When you have an Internet-facing site various search engines will be crawling your site. As you change things certain URLs might no longer work and the search engines can start generating a lot of ‘not found’ errors in your application by hitting all the links that used to exist.
+
+
+## Instead of…
+
+...getting a bunch of unhelpful, distracting noise in your monitoring setup, or errors in your logs, when Google (or another web crawler) hits deleted public pages…
+
+
+### Or...
+
+...naively swallowing all your 404 errors.
+
+```ruby
+class ApplicationController < ActionController::Base
+  rescue_from ActiveRecord::RecordNotFound, with: :not_found
+
+  private
+
+  def not_found(exception)
+    render file:   Rails.root.join('public', '404.html'),
+           layout: nil,
+           status: :not_found
+  end
+end
+```
+
+
+## Use…
+
+...the [`is_crawler`](https://github.com/ccashwell/is_crawler) gem and the configure it like so:
+
+### `application_controller.rb`
+
+```ruby
+class ApplicationController < ActionController::Base
+  include IsCrawler
+
+  rescue_from ActiveRecord::RecordNotFound, with: :not_found
+
+  private
+
+  def not_found(exception)
+    if Rails.env.production? && is_crawler?(request.user_agent)
+      render_404
+    else
+      raise exception
+    end
+  end
+
+  def render_404
+    render file:   Rails.root.join('public', '404.html'),
+           layout: nil,
+           status: :not_found
+  end
+end
+```
+
+One of my sites gets hit by many more crawlers, that aren't included by default in the gem. So I add to the list of crawlers in an initializer:
+
+### `config/initializers/is_crawler.rb`
+
+```ruby
+{
+  apple: 'Applebot',
+  arefs: 'AhrefsBot',
+  blexbot: 'BLEXBot',
+  dotbot: 'DotBot',
+  mailru: 'Mail.RU_Bot',
+  magestic12: 'MJ12bot',
+  seznam: 'SeznamBot'  
+}.each |internal_name, agent_string| do
+  Crawler::CUSTOM << Crawler.new(internal_name, agent_string)
+end
+```
+
+
+## But why?
+
+If you’re trying to make your app easier to maintain, it’s important to stay on top of your errors. You have probably wired up your application to an error monitoring tool similar to [Rollbar](https://rollbar.com), [Honeybadger](https://honeybadger.io), [Bugsnag](https://bugsnag.com), or [Sentry](https://getsentry.com).
+
+It is possible to just ‘swallow’ or ignore all 404 errors. However you want the know when real users receive 404 pages, as it might indicate something important is broken.
+
+If you cannot distinguish between the genuine issues that your visitors are having and the ‘noise’ from search engines you cannot focus and fix _real_ problems.
+
+Eradicating these ‘robot errors’ should make fixing user-facing errors more straightforward. It's good to close and fix issues as they show up in your monitoring setup, but that’s another article!
```

Comment:
I think take this paragraph out. You already explain why fixing real issues is easier without the noise above...then no point with the second sentence since you immediately say it's for another article.

Reply:
Thought the same, rearranging the paragraph made this clearer.